crates/biome_js_analyze/src/analyzers/nursery/use_node_import_protocol.rs

```diff
@@ -0,0 +1,133 @@
+use biome_analyze::{context::RuleContext, declare_rule, Ast, Rule, RuleDiagnostic};
+use biome_console::markup;
+use biome_js_syntax::JsModuleSource;
+use biome_rowan::{AstNode, TextRange};
+
+declare_rule! {
+    /// Enforces using the `node:` protocol for Node.js builtin modules.
+    ///
+    /// The prefer-node-protocol rule in ESLint enforces the use of the node: protocol
+    /// when importing Node.js builtin modules in JavaScript code.
+    /// This helps differentiate between built-in modules and third-party ones, improving code clarity.
+    /// The rule marks traditional imports like import fs from 'fs'; as invalid,
+    /// suggesting the format import fs from 'node:fs'; instead.
+    ///
+    /// Source: https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-node-protocol.md
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// import fs from 'fs';
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// import os from 'os';
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// import path from 'path';
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// import fs from 'node:fs';
+    ///
+    /// import os from 'node:os';
+    ///
+    /// import path from 'node:path';
+    /// ```
+    ///
+    pub(crate) UseNodeImportProtocol {
+        version: "next",
+        name: "useNodeImportProtocol",
+        recommended: false,
+    }
+}
+
+pub(crate) struct NodeImportProtocolState {
+    /// The range of the import statement in the source code.
+    range: TextRange,
+
+    /// The builtin module name that was imported.
+    module_name: String,
+}
+
+impl Rule for UseNodeImportProtocol {
+    type Query = Ast<JsModuleSource>;
+    type State = NodeImportProtocolState;
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let binding: &JsModuleSource = ctx.query();
+        let Ok(module_name) = binding.inner_string_text() else {
+            return None;
+        };
+
+        if !is_builtin_module(&module_name) {
+            return None;
+        }
+
+        Some(NodeImportProtocolState {
+            range: binding.range(),
+            module_name: module_name.to_string(),
+        })
+    }
+
+    fn diagnostic(_: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
+        let NodeImportProtocolState { range, module_name } = state;
+
+        Some(RuleDiagnostic::new(
+            rule_category!(),
+            *range,
+            markup! {
+                "Import from Node.js builtin module \""{module_name}"\" should use the \"node:\" protocol."
+            },
+        )
+        .note(markup! {
+            "Please change to \"node:"{module_name}"\"."
+        }))
+    }
+}
+
+const NODE_BUILTINS: &[&str] = &[
```

Comment:
Some advice:
- node contains also promises modules, like `fs/promises`. These should be taken into account
- If you sort alphabetically this array, you can use the binary search method instead of contains, which is faster
- you should take into account `require()` and `import()`
- there's another rule being developed - `noNodejsBuiltins` - which adds a similar array. Once your rule and the other are merged, we should have the same array in single place: src/globals/node.rs

Reply:
- I added more [node builtins](https://github.com/biomejs/biome/pull/1131/files#diff-7e43cb70df659129a98cbc96f2ab9f06f5936dd8988d30829b50bdf919c19540R71) and changed the location to `globals/node.rs` for other rules that will be added.
- I sorted the array alphabetically and changed contains to a [binary search](https://github.com/biomejs/biome/pull/1131/files#diff-3ccea7f725f28ab591e458af19749a00bd9419f029d8f155d6ce76dd52361772R89-R91).

However, I couldn't make it work for commonjs. I tried, but the JsModuleSource only worked for esm.
Can you give me a hint what I can try to make it work for cjs ?

Reply:
> However, I couldn't make it work for commonjs. I tried, but the JsModuleSource only worked for esm.
Can you give me a hint what I can try to make it work for cjs ?

In this case, you have to query multiple nodes. To do so, you can use the macro `declare_node_union!`. We have many rules that do so, you can take inspiration from them.  You'll have to query the node [`JsCallExpression`](https://biomejs.dev/playground/?code=cgBlAHEAdQBpAHIAZQAoACIAZgBzACIAKQA%3D), and check that `callee` is a `"require"` and it has one argument, which belongs to the node built-ins

Reply:
Thanks! I'll try it as you said.